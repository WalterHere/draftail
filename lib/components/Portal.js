import React from 'react';
import ReactDOM from 'react-dom';

const Portal = React.createClass({
    propTypes: {
        id: React.PropTypes.string.isRequired,
        children: React.PropTypes.node,
    },

    componentDidMount() {
        const { id } = this.props;

        let portal = document.getElementById(id);
        if (!portal) {
            portal = document.createElement('div');
            portal.id = id;
            document.body.appendChild(portal);
        }

        this.portal = portal;
        this.componentDidUpdate();
    },

    componentDidUpdate() {
        const { children } = this.props;

        ReactDOM.render((
            <div {...this.props}>
                {children}
            </div>
        ), this.portal);
    },

    componentWillUnmount() {
        document.body.removeChild(this.portal);
    },

    render() {
        return null;
    },
});

export default Portal;
