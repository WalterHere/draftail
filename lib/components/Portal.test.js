import React from 'react';
import { shallow } from 'enzyme';
import Portal from '../components/Portal';

describe('Portal', () => {
    it('empty', () => {
        expect(shallow(<Portal />)).toMatchSnapshot();
    });

    it('#children', () => {
        expect(shallow(<Portal>Test!</Portal>)).toMatchSnapshot();
    });

    it('component lifecycle', () => {
        const wrapper = shallow(<Portal>Test!</Portal>);

        wrapper.instance().componentDidMount();

        expect(document.body.innerHTML).toBe(
            '<div><div data-reactroot="">Test!</div></div>',
        );

        expect(wrapper.instance().portal).toBe(document.body.children[0]);

        wrapper.instance().componentDidMount();

        wrapper.instance().componentWillUnmount();

        expect(document.body.innerHTML).toBe('');
    });
});
