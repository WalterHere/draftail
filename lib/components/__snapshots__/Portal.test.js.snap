exports[`Portal #children 1`] = `null`;

exports[`Portal empty 1`] = `null`;
